webhooks-extension/pkg/endpoints/webhook_test.go

```diff
@@ -67,1024 +103,632 @@ func TestGitHubSource(t *testing.T) {
+			PullTask:         "check-me",
+		},
 	}
-}
 
-func getEnvDefaults(r *Resource, t *testing.T) EnvDefaults {
-	httpReq := dummyHTTPRequest("GET", "http://wwww.dummy.com:8080/webhooks/defaults", nil)
-	req := dummyRestfulRequest(httpReq, "")
-	httpWriter := httptest.NewRecorder()
-	resp := dummyRestfulResponse(httpWriter)
-	r.getDefaults(req, resp)
+	r := dummyResource()
+	os.Setenv("SERVICE_ACCOUNT", "tekton-test-service-account")
 
-	defaults := EnvDefaults{}
-	err := json.NewDecoder(httpWriter.Body).Decode(&defaults)
+	el, err := r.createEventListener(hooks[0], "install-namespace", hooks[0].GitRepositoryURL[strings.LastIndex(hooks[0].GitRepositoryURL, ":")+3:])
 	if err != nil {
-		t.Errorf("Error decoding result into defaults{}: %s", err.Error())
+		t.Errorf("Error creating eventlistener: %s", err)
 	}
-	return defaults
-}
-
-func createWebhook(webhook webhook, r *Resource) (response *restful.Response) {
-	// Create the first entry
-	b, err := json.Marshal(webhook)
+	el, err = r.updateEventListener(el, hooks[1], hooks[1].GitRepositoryURL[strings.LastIndex(hooks[1].GitRepositoryURL, ":")+3:])
 	if err != nil {
-		fmt.Println(fmt.Errorf("Marshal error when creating webhook, data is: %s, error is: %s", b, err))
-		return nil
+		t.Errorf("Error updating eventlistener - first time: %s", err)
 	}
-
-	httpReq := dummyHTTPRequest("POST", "http://wwww.dummy.com:8080/webhooks/", bytes.NewBuffer(b))
-	req := dummyRestfulRequest(httpReq, "")
-	httpWriter := httptest.NewRecorder()
-	resp := dummyRestfulResponse(httpWriter)
-	r.createWebhook(req, resp)
-	return resp
-}
-
-// Check a webhook's github source against k8s
-func testGitHubSource(repo, webhookName, namespace, expectedName, expectedOwnerAndRepo, expectedGitHubAPIURL string, r *Resource, t *testing.T) {
-	fmt.Printf("testGitHubSource: %s, %s, %s, %s, %s, %s \n", repo, webhookName, namespace, expectedName, expectedOwnerAndRepo, expectedGitHubAPIURL)
-	hooksForRepo, err := r.getGitHubWebhooks(repo)
+	el, err = r.updateEventListener(el, hooks[2], hooks[2].GitRepositoryURL[strings.LastIndex(hooks[2].GitRepositoryURL, ":")+3:])
 	if err != nil {
-		t.Errorf("GitHubSource %s was not found: %s", expectedName, err.Error())
+		t.Errorf("Error updating eventlistener - second time: %s", err)
 	}
 
-	ghSrc := &ghsv1alpha1.GitHubSource{}
-	for _, hook := range hooksForRepo {
-		if hook.Name == webhookName && hook.Namespace == namespace {
-			ghSrc, err = r.EventSrcClient.SourcesV1alpha1().GitHubSources(r.Defaults.Namespace).Get(hook.GithubSource, metav1.GetOptions{})
-			if err != nil {
-				t.Errorf("GitHubSource %s was not found: %s", expectedName, err.Error())
-			}
-			if ghSrc != nil {
-				fmt.Printf("GHS: %+v \n", ghSrc)
+	// Two of the webhooks are on the same repo - therefore only one monitor trigger for these
+	if len(el.Spec.Triggers) != 8 {
+		t.Errorf("Eventlistener had %d triggers, but expected 8", len(el.Spec.Triggers))
+	} else {
+		expectedTriggers := []v1alpha1.EventListenerTrigger{}
```

Comment:
We had to look at this code for 5-10 minutes to remember/work out what it's doing. Essentially it is filtering duplicate 'monitor' occurrences out of the expected names for multiple triggers. The same code is reused below, lines 338-346. It probably deserves a comment to explain what's going on.